server/src/main/java/org/elasticsearch/action/support/master/info/ClusterInfoRequest.java

```diff
@@ -42,15 +42,19 @@ public ClusterInfoRequest() {
     public ClusterInfoRequest(StreamInput in) throws IOException {
         super(in);
         indices = in.readStringArray();
-        types = in.readStringArray();
+        if (in.getVersion().before(Version.V_8_0_0)) {
+            in.readStringArray();
```

Comment:
I think these can still be non-empty in a mixed cluster with a user-request containing types routed through a 7x node

Reply:
I don't think we support the case where a user is making typed requests in a mixed 7.x and 8.x cluster?

Reply:
True, but then I think we need to throw an exception rather than assert? Or is the idea that assertions are only there in our test infrastructure, and in production it would be ignored.  In which case maybe we should throw an exception here rather than use an assertion?

Reply:
Right, it seems like we should be throwing exceptions here (and in other requests) instead of asserts. This would allow us to 'fail fast' when a user starts a rolling upgrade to 8.x but hasn't fully switched to typeless calls.

I think we use asserts in other requests like `GraphExploreRequest` (I missed that in previous reviews!), perhaps we could update to using exceptions everywhere in a future PR?